### PR TITLE
feat(config): allow percentage value for workers option

### PR DESCRIPTION
**Pull request by syi0808**
### Description

I can't use maxWorkers option that jest supported after migrating to vitest.
So I modified it so that the percentage value can be used in maxworkers and minworkers.

https://github.com/jestjs/jest/issues/9012#issue-503013549
Although it is a feature request in jest, I would like to get the same feature in viteest.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by syi0808** [comment]
I want to add test for worker option but I found two problems.
1. Can't mock `node:os` inside bundled `vitest/node`
2. if get resolveConfig function from `package/vitest/node/config.ts` throw type error when running test

I don't think I can test the code properly if I can't mock `node:os`. Is it okay to test with cpus count without mocking?
Or is it okay to test only exception handling other than calculate logic?

**Comment by sheremet-va** [comment]
> I want to add test for worker option but I found two problems.
> 
> 1. Can't mock `node:os` inside bundled `vitest/node`
> 2. If i import resolveConfig function from `package/vitest/node/config.ts` to test, throw type error from runtime.
> 
> I don't think I can test the code properly if I can't mock `node:os`. Is it okay to test with cpus count without mocking? Or is it okay to test only exception handling other than calculate logic?

You can create a separate utility that accepts the workers number and the amount of CPU and test that function.

**Comment by syi0808** [comment]
> > I want to add test for worker option but I found two problems.
> > 
> > 1. Can't mock `node:os` inside bundled `vitest/node`
> > 2. If i import resolveConfig function from `package/vitest/node/config.ts` to test, throw type error from runtime.
> > 
> > I don't think I can test the code properly if I can't mock `node:os`. Is it okay to test with cpus count without mocking? Or is it okay to test only exception handling other than calculate logic?
> 
> You can create a separate utility that accepts the workers number and the amount of CPU and test that function.

Thanks for nice idea! I separate getWorkers function to utility and node:os mock is works. I added test now.

File: test/config/test/workers-option.test.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+test('workers percent argument should not throw error', async () => {
+  const { stderr } = await runVitest({ maxWorkers: '100%', minWorkers: '10%' })
+
+  expect(stderr).toBe('')
+})
+
+describe('workers util', () => {
+  vi.mock('node:os', async importOriginal => ({
```
Comment (sheremet-va) [inline test/config/test/workers-option.test.ts:17]: `vi.mock `is always hoisted to the top of the file so it doesn't make a sense to have it inside `describe` function, let's put it at the top to make it less confusing

Comment (syi0808) [reply test/config/test/workers-option.test.ts]: Now that I look at it, I wonder how `vi.mock` worked. It's reflected.

**Review by sheremet-va — APPROVED** [review]

File: docs/guide/cli-table.md [hunk 2]
```diff
@@ -80,8 +80,8 @@
 | `--poolOptions.vmForks.minForks <workers>` | Minimum number of processes to run tests in |
```
Comment (sheremet-va) [inline docs/guide/cli-table.md:80]: We also have these options for every pool. I think they should all inherit the same behaviour then

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: docs/guide/cli-table.md (hunk 2 above)
Comment (syi0808) [reply docs/guide/cli-table.md]: I implement percentage option every pool options and now all of workflows passed!

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/vitest/src/node/config.ts [hunk 3]
```diff
@@ -436,6 +446,48 @@ export function resolveConfig(
...
+    ['vmThreads', 'minThreads'],
+    ['vmThreads', 'maxThreads'],
+  ] as const satisfies [keyof PoolOptions, keyof ThreadsOptions][]
+
+  for (const [poolOptionKey, workerOptionKey] of poolThreadsOptions) {
+    const workerInlineOption = resolved.poolOptions?.[poolOptionKey]?.[workerOptionKey]
+
+    if (workerInlineOption) {
+      resolved.poolOptions = {
+        ...resolved.poolOptions,
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/vitest/src/node/config.ts]: I think much better. done!

File: packages/vitest/src/utils/workers.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { cpus } from 'node:os'
+
+export function getWorkersCountByPercentage(percent: string) {
+  const maxWorkersCount = cpus().length
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/vitest/src/utils/workers.ts]: oh, I didn't know that function existed. done!

**Review by sheremet-va — APPROVED** [review]

**Comment by syi0808** [comment]
test failed.. because maybe i changed node:os import method. I fix test in a minutes.

File: packages/vitest/src/utils/workers.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { availableParallelism, cpus } from 'node:os'
+
+export function getWorkersCountByPercentage(percent: string) {
+  const maxWorkersCount = availableParallelism?.() ?? cpus().length
```
Comment (sheremet-va) [inline packages/vitest/src/utils/workers.ts:4]: This needs to be on the default export otherwise Node.js will fail at runtime

Comment (syi0808) [reply packages/vitest/src/utils/workers.ts:4]: I see. I will change node:os mock code.

Comment (syi0808) [reply packages/vitest/src/utils/workers.ts]: I mock default export of node:os.

**Review by sheremet-va — APPROVED** [review]

**Comment by sheremet-va** [comment]
LGTM! Thank you! 🎉

**Pull request merged**